source/Octopus.Client/OctopusAsyncClient.cs

```diff
@@ -150,110 +144,39 @@ private static async Task<IOctopusAsyncClient> Create(OctopusServerEndpoint serv
-                client.clientOptions.SpaceContext = defaultSpace != null ? SpaceContext.SpecificSpaceAndSystem(defaultSpace.Id) : SpaceContext.SystemOnly();
-            }
-
-            var spaceId = GetSpaceId(client.SpaceContext);
-            return !string.IsNullOrEmpty(spaceId) ?
-                await client.Get<SpaceRootResource>(rootResource.Link("SpaceHome"), new { spaceId }).ConfigureAwait(false)
-                : null;
-        }
-
-
-        /// <summary>
-        /// Gets a document that identifies the Octopus server (from /api) and provides links to the resources available on the
-        /// server. Instead of hardcoding paths,
-        /// clients should use these link properties to traverse the resources on the server. This document is lazily loaded so
-        /// that it is only requested once for
-        /// the current <see cref="IOctopusAsyncClient" />.
-        /// </summary>
-        public RootResource RootDocument => RootDocuments.RootResource;
-
-        public SpaceRootResource SpaceRootDocument => RootDocuments.SpaceRootResource;
-
-        RootResources RootDocuments { get; set; }
-
-        class RootResources
-        {
-            public RootResources(RootResource rootResource, SpaceRootResource spaceRootResource)
-            {
-                RootResource = rootResource;
-                SpaceRootResource = spaceRootResource;
-            }
-
-            public RootResource RootResource { get; }
-            public SpaceRootResource SpaceRootResource { get; }
-        }
+        [Obsolete("This property is deprecated, please use the one from Repository instead")]
+        public RootResource RootDocument => Repository?.RootDocument;
 
         /// <summary>
         /// Indicates whether a secure (SSL) connection is being used to communicate with the server.
         /// </summary>
         public bool IsUsingSecureConnection => serverEndpoint.IsUsingSecureConnection;
 
-        /// <summary>
-        /// Requests a fresh root document from the Octopus Server which can be useful if the API surface has changed. This can occur when enabling/disabling features, or changing license.
-        /// </summary>
-        /// <returns>A fresh copy of the root document.</returns>
+        [Obsolete("This method is deprecated, please use the one from Repository instead")]
         public async Task<RootResource> RefreshRootDocument()
         {
-            var rootDocument = await Get<RootResource>(rootDocumentUri).ConfigureAwait(false);
-            var spaceRootDocument = await LoadSpaceRootResource(this, rootDocument);
-            RootDocuments = new RootResources(rootDocument, spaceRootDocument);
-            return rootDocument;
+            return await Repository.RefreshRootDocument();
         }
 
-        public async Task<IOctopusAsyncClient> ForSpace(string spaceId)
+        protected virtual Task<RootResource> EstablishSession()
```

Comment:
Do you know why this guy is protected virtual? Looks pretty weird and this is a kind of breaking change.

Reply:
Not too sure, I cannot find the usage of it, but I might not know enough